src/compiler/crystal/command.cr

```diff
@@ -364,6 +364,23 @@ class Crystal::Command
         opts.on("--mattr CPU", "Target specific features") do |features|
           compiler.mattr = features
         end
+        opts.on("--mcmodel CPU", "Target specific code model") do |mcmodel|
+          compiler.mcmodel = case mcmodel
+                             when "default"
+                               LLVM::CodeModel::Default
+                             when "small"
+                               LLVM::CodeModel::Small
+                             when "kernel"
+                               LLVM::CodeModel::Kernel
+                             when "medium"
+                               LLVM::CodeModel::Medium
+                             when "large"
+                               LLVM::CodeModel::Large
+                             else
+                               error "--mcmodel should be one of: default, kernel, small, medium, large"
+                               raise "unreachable"
```

Comment:
```suggestion
```
Does it not compile without this? Isn't `error` a `NoReturn` method? I think so because then this shouldn't be required.

Reply:
It's required or the compiler will spill out:

```
Error: instance variable '@mcmodel' of Crystal::Compiler must be LLVM::CodeModel, not (LLVM::CodeModel | Nil)
```

Reply:
```suggestion
                               exit 1
```

Reply:
`#error` doesn't exit. You need to call `.exit` explicitly.

Reply:
It does for me; Do I really need to do it?

Reply:
Weird. This might be a compiler bug.

Reply:
Yes. `raise "unreachable"` is incorrect because this location is very much reachable as soon as you use an incorrect flag value. At this point there should not be an exception raised. The command should just exit with an error.

Reply:
This is strange because this is the method called:
https://github.com/crystal-lang/crystal/blob/9f34c313d3cb3833565623a8f18a115bac1544f8/src/compiler/crystal/command.cr#L576
and in this method this method is called:
https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/util.cr#L14
and in this method the top-level method `exit` is called. Which is a `NoReturn` method so there shouldn't be a `raise` or another `exit 1` required after this.

Reply:
@r00ster91 you're right. I don't understand why this wouldn't work.
This isolated example is just fine and as expected: https://carc.in/#/r/7upy

Reply:
The compiler might not know that something is unreachable, and eventually find it out, but then it might be too late and the nil will be there. So yes, probably a compiler bug.

Reply:
The problem here is that `error` will do a [conditional](https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/util.cr#L17) exit. So the return type is `NoReturn | Nil`.

Since the behavior is to actually exit based on the runtime values that will always be there a `raise "unreachable"` is more accurate IMO.

I prefer to leave the `raise "unreachable"` here and also undo the following change in line 541/553 that replace the `raise "unreachable"` with an `exit 1` for the `--warnings`

Reply:
An alternative solution would be to refactor `error` into separate methods which are dispatched depending on the type of `exit_code`.

```cr
  def self.error(msg, color, exit_code : Nil, stderr = STDERR, leading_error = true) : Nil
    stderr.print "Error: ".colorize.toggle(color).red.bold if leading_error
    stderr.puts msg.colorize.toggle(color).bright
  end

  def self.error(msg, color, exit_code : Int = 1, stderr = STDERR, leading_error = true)
    error(msg, color, nil, stderr, leading_error)
    exit(exit_code)
  end
```

Reply:
Yes, or have something `error_and_exit`/`fatal_error`/`cli_option_error`. But that is for another PR. My comments apply for accepting this PR in the current state.